File: src/Application Specific/Classes/Entity Collections/Order/OrderConsolidation/OrderConsolidationValidator.cs
using System.Collections;
using System.Data;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace FSM.Entity
{
    namespace OrderConsolidations
    {
        // make sure that contact object is valid
        public class OrderConsolidationValidator : BaseValidator
        {
            public void Validate(OrderConsolidation oOrderConsolidation, bool ForLocation)
            {

                // make sure that contact object is valid
                if (oOrderConsolidation.Errors.Count > 0)
                {
                    DictionaryEntry de;
                    foreach (var de in oOrderConsolidation.Errors)
                        AddCriticalMessage(Conversions.ToString(de.Value));
                }

                if (ForLocation)
                {
                    if (oOrderConsolidation.LocationID == 0)
                    {
                        AddCriticalMessage("Location Missing");
                    }
                }
                else if (oOrderConsolidation.SupplierID == 0)
                {
                    AddCriticalMessage("Supplier Missing");
                }

                if (oOrderConsolidation.ConsolidationRef.Trim().Length == 0)
                {
                    AddCriticalMessage("Reference Missing");
                }
                else if (!oOrderConsolidation.Exists)
                {
                    if (App.DB.Order.Order_CheckReference(oOrderConsolidation.ConsolidationRef).Table.Rows.Count > 0)
                    {
                        AddCriticalMessage("The Order Ref you have chosen has already been allocated to another Quote or Order in the system. Please choose a unique Reference.");
                    }
                }

                var orders = new ArrayList();
                if (oOrderConsolidation.HasSupplierPO)
                {
                    foreach (DataRow row in App.DB.Order.Order_GetAll("").Table.Rows)
                    {
                        if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(row["OrderConsolidationID"], oOrderConsolidation.OrderConsolidationID, false)))
                        {
                            if (Conversions.ToBoolean(row["ReadyToSendToSage"]))
                            {
                                orders.Add(row["OrderReference"]);
                            }
                        }
                    }
                }

                if (!(orders.Count == 0))
                {
                    string msg = "The following orders have been marked as ready to send to sage so this consolidation cannot be marked with a supplier PO:" + Constants.vbCrLf + Constants.vbCrLf;
                    foreach (string Err in orders)
                        msg += Err + Constants.vbCrLf;
                    AddCriticalMessage(msg);
                }

                if (oOrderConsolidation.ReadyToSendToSage)
                {
                    if (oOrderConsolidation.SupplierInvoiceReference.Trim().Length == 0)
                    {
                        AddCriticalMessage("Supplier Invoice Reference Missing");
                    }

                    if (oOrderConsolidation.SupplierInvoiceDate == default)
                    {
                        AddCriticalMessage("Supplier Invoice Date Missing");
                    }

                    if (oOrderConsolidation.NominalCode.Trim().Length == 0)
                    {
                        AddCriticalMessage("Nominal Code Missing");
                    }

                    if (oOrderConsolidation.DepartmentRef.Trim().Length == 0)
                    {
                        AddCriticalMessage("Department Reference Missing");
                    }

                    if (oOrderConsolidation.ExtraRef.Trim().Length == 0)
                    {
                        AddCriticalMessage("Extra Reference Missing");
                    }

                    if (oOrderConsolidation.TaxCodeID == 0)
                    {
                        AddCriticalMessage("Tax Code Missing");
                    }
                }

                if (ValidatorMessages.CriticalMessages.Count > 0)
                {
                    throw new ValidationException(this);
                }
            }
        }
    }
}

File: src/Application Specific/Classes/Entity Collections/Order/OrderConsolidation/OrderConsolidationValidator.vb
Namespace Entity

    Namespace OrderConsolidations

        Public Class OrderConsolidationValidator
            'make sure that contact object is valid
            Inherits BaseValidator

            Public Sub Validate(ByVal oOrderConsolidation As OrderConsolidation, ByVal ForLocation As Boolean)

                'make sure that contact object is valid
                If oOrderConsolidation.Errors.Count > 0 Then
                    Dim de As DictionaryEntry
                    For Each de In oOrderConsolidation.Errors
                        Me.AddCriticalMessage(CStr(de.Value))
                    Next
                End If

                If ForLocation Then
                    If oOrderConsolidation.LocationID = 0 Then
                        Me.AddCriticalMessage("Location Missing")
                    End If
                Else
                    If oOrderConsolidation.SupplierID = 0 Then
                        Me.AddCriticalMessage("Supplier Missing")
                    End If
                End If
                If oOrderConsolidation.ConsolidationRef.Trim.Length = 0 Then
                    Me.AddCriticalMessage("Reference Missing")
                Else
                    If Not oOrderConsolidation.Exists Then
                        If DB.Order.Order_CheckReference(oOrderConsolidation.ConsolidationRef).Table.Rows.Count > 0 Then
                            Me.AddCriticalMessage("The Order Ref you have chosen has already been allocated to another Quote or Order in the system. Please choose a unique Reference.")
                        End If
                    End If
                End If

                Dim orders As New ArrayList
                If oOrderConsolidation.HasSupplierPO Then
                    For Each row As DataRow In DB.Order.Order_GetAll("").Table.Rows
                        If row.Item("OrderConsolidationID") = oOrderConsolidation.OrderConsolidationID Then
                            If row.Item("ReadyToSendToSage") Then
                                orders.Add(row.Item("OrderReference"))
                            End If
                        End If
                    Next
                End If

                If Not orders.Count = 0 Then
                    Dim msg As String = "The following orders have been marked as ready to send to sage so this consolidation cannot be marked with a supplier PO:" & vbCrLf & vbCrLf

                    For Each Err As String In orders
                        msg += Err & vbCrLf
                    Next

                    Me.AddCriticalMessage(msg)
                End If

                If oOrderConsolidation.ReadyToSendToSage Then
                    If oOrderConsolidation.SupplierInvoiceReference.Trim.Length = 0 Then
                        Me.AddCriticalMessage("Supplier Invoice Reference Missing")
                    End If
                    If oOrderConsolidation.SupplierInvoiceDate = Nothing Then
                        Me.AddCriticalMessage("Supplier Invoice Date Missing")
                    End If
                    If oOrderConsolidation.NominalCode.Trim.Length = 0 Then
                        Me.AddCriticalMessage("Nominal Code Missing")
                    End If
                    If oOrderConsolidation.DepartmentRef.Trim.Length = 0 Then
                        Me.AddCriticalMessage("Department Reference Missing")
                    End If
                    If oOrderConsolidation.ExtraRef.Trim.Length = 0 Then
                        Me.AddCriticalMessage("Extra Reference Missing")
                    End If
                    If oOrderConsolidation.TaxCodeID = 0 Then
                        Me.AddCriticalMessage("Tax Code Missing")
                    End If
                End If

                If Me.ValidatorMessages.CriticalMessages.Count > 0 Then
                    Throw New ValidationException(Me)
                End If

            End Sub

        End Class

    End Namespace

End Namespace
